### PR TITLE
Avoid ptr int round trips

### PR DESCRIPTION
**Pull request by MitchBriles**
Replace instances of `ptrToInt`->`add`->`intToPtr` with a single `getelementptr` instruction—possible performance benefits.

Needs testing! Some examples of it working in [this repo](https://github.com/MitchBriles/PtrIntRoundTripRemover).

### CODE REVIEW

**Comment by regehr** [comment]
ok, awesome! 
I'll leave some comments on this PR, but in the meantime, please update it to include at least one test case. you can just port one of them over from your other repo. it would be best to include at least one positive test case (indicating that the desired transformation happens when it's supposed to) and also at least one negative test case (indicating that the desired transformation doesn't happen when it's not supposed to).

File: backend_tv/adjust_src.cpp [hunk 1]
```diff
@@ -481,4 +481,43 @@ void fixupOptimizedTgt(llvm::Function *tgt) {
     myAllocCall->eraseFromParent();
 }
 
+/**
+ * Given an intToPtr instruction, checks for a round trip 
+ * from a ptrToInt instruction, then replaces with a single GEP instruction.
+ * This translation can be verified with physical pointers.
+ */
+void tryReplaceRoundTrip(llvm::IntToPtrInst *intToPtr) {
+  if (!intToPtr)
```
Comment (regehr) [inline backend_tv/adjust_src.cpp:490]: you don't need to do this check if you've already tested this pointer for non-null elsewhere (as you have done). if you don't feel comfortable totally removing this check, you can replace it with an assertion.

File: backend_tv/adjust_src.cpp [hunk 2]
```diff
@@ -481,4 +481,43 @@ void fixupOptimizedTgt(llvm::Function *tgt) {
...
+    return;
+
+  // we only understand add instructions in this context
+  auto *op_inst = dyn_cast<llvm::Instruction>(intToPtr->getOperand(0));
+  if (!op_inst || op_inst->getOpcode() != llvm::Instruction::Add || op_inst->getNumUses() > 1)
+      return;
+
+  // keep track of (ptr + int) vs (int + ptr)
+  bool ptrOnLeft = true;
```
Comment (regehr) [inline backend_tv/adjust_src.cpp:499]: ok, since you have two distinct cases in your optimization, it would be best to add two test cases, one corresponding to each thing that can happen here

File: tools/backend-tv.cpp [hunk 3]
```diff
@@ -218,6 +219,16 @@ void doit(llvm::Module *srcModule, llvm::Function *srcFn, Verifier &verifier,
 
   lifter::fixupOptimizedTgt(F2);
 
+  // find and collapse sequences of the form ptrToInt, add, intToPtr
+  // into a single GEP instruction. Possible performance benefits.
+  // apparently iterating in this way is stable
+  for (auto it = instructions(*F2).begin(), end = instructions(*F2).end(); it != end;) {
+      llvm::Instruction& Inst = *it++;
```
Comment (regehr) [inline tools/backend-tv.cpp:226]: can you check again if there's a more graceful way to do this than having "it++" outside of the loop construct? I feel like there must be

Comment (MitchBriles) [reply tools/backend-tv.cpp:226]: I tried messing around with it and can't find a nice way to do it. I found that I can inline it like this:
``` 
if (auto *intToPtr = dyn_cast<IntToPtrInst>(&*it++)) {
```
But I don't think this is very clear. Please let me know if you find a better way.

**Comment by MitchBriles** [comment]
> hi Mitch, this PR includes a lot of unrelated changes, can you rebase it so that it only contains your specific changes? with all of the noise in this patch I can't easily tell which is code you wrote

Oops you're right. Rebased it.
